File: bubble_sort.js
function bubbleSort(arr) {
    /**
     * Bubble Sort Algorithm
     * Time Complexity: O(n²)
     * Space Complexity: O(1)
     */
    const n = arr.length;
    const steps = [];
    
    // Create a copy to avoid modifying original
    const arrCopy = [...arr];
    
    for (let i = 0; i < n; i++) {
        let swapped = false;
        for (let j = 0; j < n - i - 1; j++) {
            // Record comparison step
            steps.push({
                type: 'compare',
                indices: [j, j + 1],
                array: [...arrCopy],
                message: `Comparing elements at indices ${j} and ${j + 1}`
            });
            
            if (arrCopy[j] > arrCopy[j + 1]) {
                // Record swap step
                [arrCopy[j], arrCopy[j + 1]] = [arrCopy[j + 1], arrCopy[j]];
                swapped = true;
                steps.push({
                    type: 'swap',
                    indices: [j, j + 1],
                    array: [...arrCopy],
                    message: `Swapped elements at indices ${j} and ${j + 1}`
                });
            }
        }
        
        // Record completion of pass
        steps.push({
            type: 'pass_complete',
            indices: [n - i - 1],
            array: [...arrCopy],
            message: `Pass ${i + 1} complete. Element at index ${n - i - 1} is in final position`
        });
        
        if (!swapped) {
            break;
        }
    }
    
    return steps;
}

// Example usage
if (typeof module !== 'undefined' && module.exports) {
    module.exports = bubbleSort;
} else {
    // Browser environment
    window.bubbleSort = bubbleSort;
}

// Test example
const testArray = [64, 34, 25, 12, 22, 11, 90];
console.log(`Original array: ${testArray}`);
const steps = bubbleSort(testArray);
console.log(`Number of steps: ${steps.length}`);

File: bubble_sort.py
def bubble_sort(arr):
    """
    Bubble Sort Algorithm
    Time Complexity: O(n²)
    Space Complexity: O(1)
    """
    n = len(arr)
    steps = []
    
    # Create a copy to avoid modifying original
    arr_copy = arr.copy()
    
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            # Record comparison step
            steps.append({
                'type': 'compare',
                'indices': [j, j + 1],
                'array': arr_copy.copy(),
                'message': f'Comparing elements at indices {j} and {j + 1}'
            })
            
            if arr_copy[j] > arr_copy[j + 1]:
                # Record swap step
                arr_copy[j], arr_copy[j + 1] = arr_copy[j + 1], arr_copy[j]
                swapped = True
                steps.append({
                    'type': 'swap',
                    'indices': [j, j + 1],
                    'array': arr_copy.copy(),
                    'message': f'Swapped elements at indices {j} and {j + 1}'
                })
        
        # Record completion of pass
        steps.append({
            'type': 'pass_complete',
            'indices': [n - i - 1],
            'array': arr_copy.copy(),
            'message': f'Pass {i + 1} complete. Element at index {n - i - 1} is in final position'
        })
        
        if not swapped:
            break
    
    return steps

# Example usage
if __name__ == "__main__":
    test_array = [64, 34, 25, 12, 22, 11, 90]
    steps = bubble_sort(test_array)
    print(f"Original array: {test_array}")
    print(f"Number of steps: {len(steps)}")

